common/diagnostics/StatisticReporter.java

```diff
@@ -10,49 +10,117 @@
-    protected static final Logger LOG = LoggerFactory.getLogger(StatisticReporter.class);
+    // Modify calculateInitialDelay() if you change this value!
+    protected static final int REPORT_INTERVAL_MINUTES = 60;
+    public static final String DISABLED_REPORTING_FILE_NAME = "_reporting_disabled";
 
-    private final String reportingURI;
+    protected static final Logger LOG = LoggerFactory.getLogger(StatisticReporter.class);
+    private final String deploymentID;
     private final Metrics metrics;
+    private final boolean statisticsReportingEnabled;
+    private final String reportingURI;
+    private final Path dataDirectory;
 
-    private ScheduledFuture<?> pushScheduledTask;
+    private ScheduledFuture<?> reportScheduledTask;
     private final ScheduledThreadPoolExecutor scheduled = new ScheduledThreadPoolExecutor(1);
 
-    public StatisticReporter(Metrics metrics, String reportingURI) {
+    public StatisticReporter(String deploymentID, Metrics metrics, boolean statisticsReportingEnabled, String reportingURI, Path dataDirectory) {
+        this.deploymentID = deploymentID;
         this.metrics = metrics;
+        this.statisticsReportingEnabled = statisticsReportingEnabled;
         this.reportingURI = reportingURI;
-        push();
+        this.dataDirectory = dataDirectory;
+    }
+
+    public void startReporting() {
+        if (statisticsReportingEnabled) {
+            deleteDisabledReportingFileIfExists();
+            scheduleReporting();
+        } else {
+            reportOnceIfNeeded();
+        }
     }
 
-    private void push() {
+    private boolean report() {
         try {
-            HttpsURLConnection conn = (HttpsURLConnection)(new URL(reportingURI)).openConnection();
+            String metricsJSON = metrics.formatJSON(true);
+            metrics.takeSnapshot(); // Forget about this period for consistency even if the https write is not successful
 
-            conn.setRequestMethod("POST");
+            HttpsURLConnection conn = (HttpsURLConnection) (new URL(reportingURI)).openConnection();
 
+            conn.setRequestMethod("POST");
             conn.setRequestProperty("Charset", "utf-8");
             conn.setRequestProperty("Connection", "close");
             conn.setRequestProperty("Content-Type", "application/json");
 
             conn.setDoOutput(true);
-            conn.getOutputStream().write(metrics.formatJSON().getBytes(StandardCharsets.UTF_8));
+            conn.getOutputStream().write(metricsJSON.getBytes(StandardCharsets.UTF_8));
 
             conn.connect();
-
             conn.getInputStream().readAllBytes();
+            return true;
         } catch (Exception e) {
-            if (LOG.isTraceEnabled()) LOG.trace("Failed to push metrics to {}:", reportingURI, e);
-            // do nothing
-        } finally {
-            pushScheduledTask = scheduled.schedule(this::push, 1, HOURS);
+            LOG.warn("Failed to push metrics to {}: {}", reportingURI, e.getMessage());
```

Comment:
@dmitrii-ubskii I think that it's a good thing to log as the diagnostics server is expected to be online most of the time. In the worst case, it's 24 lines of log entries per day. Should not be bad.

Reply:
Well... You're right, and since it already works, I'll return the trace, I guess.